From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rob Norris <rob.norris@truenas.com>
Date: Sun, 16 Aug 2026 09:14:34 +1000
Subject: [PATCH 3/5] Linux 7.3: inode_operations.create dropped flag arg

Sponsored-by: TrueNAS
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Tony Hutter <hutter2@llnl.gov>
Co-developed-by: TheAlmightyOgreLord <beardedtechwizard@protonmail.com>
Signed-off-by: Rob Norris <rob.norris@truenas.com>
Closes #19082

Backported to 2.4.4: folded into the existing ZFS_AC_KERNEL_CREATE
check instead of a separate m4 file (upstream commit 6e2fa4bc87c3).

Signed-off-by: x <a@b>
---
 config/kernel-inode-create.m4   | 28 ++++++++++++++++++++++++++++
 module/os/linux/zfs/zpl_inode.c |  3 +++
 2 files changed, 31 insertions(+)

diff --git a/config/kernel-inode-create.m4 b/config/kernel-inode-create.m4
index 65817ca..65e8f44 100644
--- a/config/kernel-inode-create.m4
+++ b/config/kernel-inode-create.m4
@@ -1,5 +1,23 @@
 dnl # SPDX-License-Identifier: CDDL-1.0
 AC_DEFUN([ZFS_AC_KERNEL_SRC_CREATE], [
+	dnl #
+	dnl # 7.3 API change
+	dnl # The trailing bool flag arg was dropped
+	dnl #
+	ZFS_LINUX_TEST_SRC([create_mnt_idmap_no_flag], [
+		#include <linux/fs.h>
+		#include <linux/sched.h>
+
+		static int inode_create(struct mnt_idmap *idmap,
+		    struct inode *inode ,struct dentry *dentry,
+		    umode_t umode) { return 0; }
+
+		static const struct inode_operations
+			iops __attribute__ ((unused)) = {
+			.create         = inode_create,
+		};
+	],[])
+
 	dnl #
 	dnl # 6.3 API change
 	dnl # The first arg is changed to struct mnt_idmap *
@@ -54,6 +72,15 @@ AC_DEFUN([ZFS_AC_KERNEL_SRC_CREATE], [
 ])
 
 AC_DEFUN([ZFS_AC_KERNEL_CREATE], [
+	AC_MSG_CHECKING([whether iops->create() takes a flag arg])
+	ZFS_LINUX_TEST_RESULT([create_mnt_idmap_no_flag], [
+		AC_MSG_RESULT(no)
+		AC_DEFINE(HAVE_IOPS_CREATE_IDMAP, 1,
+		   [iops->create() takes struct mnt_idmap*])
+		AC_DEFINE(HAVE_IOPS_CREATE_NO_FLAG_ARG, 1,
+		   [iops->create() does not take a flag arg])
+	],[
+	AC_MSG_RESULT(yes)
 	AC_MSG_CHECKING([whether iops->create() takes struct mnt_idmap*])
 	ZFS_LINUX_TEST_RESULT([create_mnt_idmap], [
 		AC_MSG_RESULT(yes)
@@ -78,4 +105,5 @@ AC_DEFUN([ZFS_AC_KERNEL_CREATE], [
 			])
 		])
 	])
+	])
 ])
diff --git a/module/os/linux/zfs/zpl_inode.c b/module/os/linux/zfs/zpl_inode.c
index 3492f4a..1157378 100644
--- a/module/os/linux/zfs/zpl_inode.c
+++ b/module/os/linux/zfs/zpl_inode.c
@@ -168,6 +168,9 @@ static int
 #ifdef HAVE_IOPS_CREATE_USERNS
 zpl_create(struct user_namespace *user_ns, struct inode *dir,
     struct dentry *dentry, umode_t mode, bool flag)
+#elif defined(HAVE_IOPS_CREATE_NO_FLAG_ARG)
+zpl_create(struct mnt_idmap *user_ns, struct inode *dir,
+    struct dentry *dentry, umode_t mode)
 #elif defined(HAVE_IOPS_CREATE_IDMAP)
 zpl_create(struct mnt_idmap *user_ns, struct inode *dir,
     struct dentry *dentry, umode_t mode, bool flag)
